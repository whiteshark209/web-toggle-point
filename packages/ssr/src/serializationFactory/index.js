import jsesc from "jsesc";
import parse from "html-react-parser";

// eslint-disable-next-line prettier/prettier, no-empty -- https://github.com/babel/babel/issues/15156
{}
/**
 * A factory for creating a serialization object that has methods for serializing and deserializing JSON data in server-rendered web applications.
 * @memberof module:asos-web-toggle-point-ssr
 * @inner
 * @function
 * @memberof module:asos-web-toggle-point-ssr
 * @inner
 * @function
 * @param {object} params parameters
 * @param {string} params.id The id attribute of the backing application/json script.
 * @param {external:HostApplication.logWarning} params.logWarning A method that logs warnings; will be used when malformed JSON is found in the backing store when deserialize on the client, which should only be possible if processed in a system downstream from the origin.
 * @returns {module:asos-web-toggle-point-ssr.serialization} Some serialization / deserialization methods
 * @example
 * const logWarning = (warning) => console.log(warning);
 * const id = "app_features";
 * const { getScriptMarkup, getJSONFromScript } = serializationFactory({ id, logWarning });
 */
const serializationFactory = ({ id, logWarning }) =>
  /**
   * @typedef {function} getScriptMarkup
   * @memberof module:asos-web-toggle-point-ssr
   * @param {object} params parameters
   * @param {object} params.content The JSON content to be serialized.
   */
  /**
   * @typedef {function} getScriptReactElement
   * @memberof module:asos-web-toggle-point-ssr
   * @param {object} params parameters
   * @param {object} params.content The JSON content to be serialized.
   */
  /**
   * @typedef {function} getJSONFromScript
   * @memberof module:asos-web-toggle-point-ssr
   * @returns {object} The JSON content of the script element.
   */

  /**
   * An object containing methods for serializing and deserializing JSON data in server-rendered web applications.
   * @typedef {object} serialization
   * @memberof module:asos-web-toggle-point-ssr
   * @property {module:asos-web-toggle-point-ssr.getScriptMarkup} getScriptMarkup Gets a string containing markup for a type="application/json" script element with the specified content.
   * @property {module:asos-web-toggle-point-ssr.getScriptReactElement} getScriptReactElement - Gets a React element for a type="application/json" script element with the specified content.
   * @property {module:asos-web-toggle-point-ssr.getJSONFromScript} getJSONFromScript - Returns the JSON content of the script element.
   */
  ({
    /**
     * @memberof module:asos-web-toggle-point-ssr.serialization
     * @param {object} content The JSON content to be serialized.
     * @returns {string} A string containing markup for a type="application/json" script element with the specified content.
     */
    getScriptMarkup({ content }) {
      return `<script id="${id}" type="application/json">${jsesc(content, {
        isScriptContext: true,
        json: true
      })}</script>`;
    },
    getScriptReactElement(...args) {
      return parse(this.getScriptMarkup(...args));
    },
    getJSONFromScript() {
      const input = document.querySelector(`#${id}`)?.textContent;
      let output = null;
      try {
        output = JSON.parse(input);
      } catch {
        logWarning(`Invalid JSON found in markup: "${input}"`);
      }

      return output;
    }
  });

export default serializationFactory;
