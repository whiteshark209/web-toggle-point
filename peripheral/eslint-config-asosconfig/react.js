import browser from "./browser.js";
import reactPlugin from "eslint-plugin-react";
import jsxA11y from "eslint-plugin-jsx-a11y";
import { fixupPluginRules } from "@eslint/compat";
import reactHooks from "eslint-plugin-react-hooks";

export default [
  ...browser,
  {
    ...reactPlugin.configs.flat.recommended
  },
  {
    ...reactPlugin.configs.flat["jsx-runtime"]
  },
  {
    ...jsxA11y.flatConfigs.recommended
  },
  {
    name: "asosconfig/react",
    languageOptions: {
      parserOptions: {
        ecmaFeatures: {
          jsx: true
        }
      }
    },
    plugins: {
      "react-hooks": fixupPluginRules(reactHooks)
    },
    rules: {
      ...reactHooks.configs.recommended.rules,
      "jsx-a11y/aria-props": "warn",
      "jsx-a11y/label-has-for": [
        2,
        {
          required: {
            every: ["id"]
          }
        }
      ]
    }
  }
];
