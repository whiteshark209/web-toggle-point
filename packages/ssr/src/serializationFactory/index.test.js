import serializationFactory from "./index";
import jsesc from "jsesc";
import { render } from "@testing-library/react";

jest.mock("jsesc", () => jest.fn((x) => `secure: ${JSON.stringify(x)}`));

describe("serializationFactory", () => {
  const logWarning = jest.fn();
  const id = "test-id";
  const content = { test: "value" };
  let serialization;

  beforeEach(() => {
    logWarning.mockClear();
    serialization = serializationFactory({ id, logWarning });
  });

  it("should return an object with three methods", () => {
    expect(serialization).toMatchObject({
      getScriptMarkup: expect.any(Function),
      getScriptReactElement: expect.any(Function),
      getJSONFromScript: expect.any(Function)
    });
  });

  describe("when getting script markup", () => {
    let result;

    beforeEach(() => {
      result = serialization.getScriptMarkup({ content });
    });

    it("should return a string containing a JSON script tag, appropriately quoted by jsesc", () => {
      expect(jsesc).toHaveBeenCalledWith(content, {
        isScriptContext: true,
        json: true
      });
      expect(result).toStrictEqual(
        `<script id="${id}" type="application/json">${jsesc(content)}</script>`
      );
    });
  });

  describe("when rendering a scriptReactElement with React", () => {
    let result;
    beforeEach(() => {
      ({ container: result } = render(
        serialization.getScriptReactElement({ content })
      ));
    });

    it("should return a react element containing a JSON script tag, appropriately quoted by jsesc", () => {
      expect(result.firstChild).toMatchSnapshot();
    });
  });

  describe("when getting JSON from script", () => {
    let result;
    const validJsonString =
      '{"test-experiment-a":{"test-prop-1":"test-value-1"},"test-experiment-b":{"test-prop-2":"test-value-2"}}';

    const invalidJsonString =
      '{"test-experiment-a":{"test-prop-1":"test-value-1"},"test-experiment-b":{"test-prop-2":"test-value-2"},}';

    const jsonObject = {
      "test-experiment-a": { "test-prop-1": "test-value-1" },
      "test-experiment-b": { "test-prop-2": "test-value-2" }
    };

    describe("when a valid JSON string is passed", () => {
      beforeEach(() => {
        document.body.innerHTML = `<script id="${id}" type="application/json">${validJsonString}</script>`;
        result = serialization.getJSONFromScript();
      });

      afterEach(() => {
        document.body.innerHTML = "";
      });

      it("should return a valid object", () => {
        expect(result).toStrictEqual(jsonObject);
      });

      it("should not have called any logger warning", () => {
        expect(logWarning).not.toHaveBeenCalled();
      });
    });

    describe("when an invalid JSON string is passed", () => {
      beforeEach(() => {
        document.body.innerHTML = `<script id="${id}" type="application/json">${invalidJsonString}</script>`;
        result = serialization.getJSONFromScript();
      });

      it("should have logged a warning", () => {
        expect(logWarning).toHaveBeenCalledWith(
          `Invalid JSON found in markup: "${invalidJsonString}"`
        );
      });

      it("should return null", () => {
        expect(result).toStrictEqual(null);
      });
    });
  });
});
