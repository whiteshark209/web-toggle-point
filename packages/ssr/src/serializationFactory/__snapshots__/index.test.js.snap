// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serializationFactory when rendering a scriptReactElement with React should return a react element containing a JSON script tag, appropriately quoted by jsesc 1`] = `
<script
  id="test-id"
  type="application/json"
>
  secure: {"test":"value"}
</script>
`;
